test/hotspot/jtreg/sources/SortIncludes.java

```diff
@@ -0,0 +1,230 @@
+        // be put in the .hpp file, and not in the .inline.hpp file. This rule
+        // exists to resolve problems with circular dependencies between
+        // .inline.hpp files.
+        String pathString = path.toString();
+        boolean isInlineHpp = pathString.endsWith(".inline.hpp");
+        String nonInlineHpp = pathString.replace(".inline.hpp", ".hpp");
+
+        List<String> result = new ArrayList<>(lines.length);
+
+        // Partition lines into user include and sys includes and discard blank lines
+        for (String line : lines) {
+            int doubleQuote = line.indexOf('"');
+            if (line.contains("\"")) {
+                if (isInlineHpp && nonInlineHpp.endsWith(extract(line, '"', '"'))) {
+                    result.add(line);
+                } else {
+                    userIncludes.add(line);
+                }
+            } else if (line.contains("<")) {
+                sysIncludes.add(line);
+            }
+        }
+
+        if (!result.isEmpty() && (!userIncludes.isEmpty() || !sysIncludes.isEmpty())) {
+            // Insert blank line between include of .hpp from .inline.hpp
+            // and the rest of the includes
+            result.add("");
+        }
+        result.addAll(userIncludes);
+        if (!userIncludes.isEmpty() && !sysIncludes.isEmpty()) {
+            // Insert blank line between user and sys includes
+            result.add("");
+        }
+        result.addAll(sysIncludes);
+
+        return String.join("\n", result) + "\n";
+    }
+
+    /// Processes the C++ source file in `path` to sort its include statements.
+    ///
+    /// @param path a path of a C++ source file
+    /// @param update updates the source file if sorting changed its content
+    /// @return `true` if sorting changes were made,`false` otherwise
+    public static boolean sortIncludes(Path path, boolean update) throws IOException {
+        String source = Files.readString(path);
+        Matcher matcher = INCLUDES_RE.matcher(source);
+        StringBuilder buf = new StringBuilder();
+        int end = 0;
+
+        while (matcher.find()) {
+            if (matcher.start() != end) {
+                buf.append(source, end, matcher.start());
+            }
+            buf.append(sortedIncludes(path, matcher.group()));
+            end = matcher.end();
+        }
+
+        if (end == 0) {
+            return false;
+        }
+        buf.append(source.substring(end));
+
+        String newSource = buf.toString();
+        if (!newSource.equals(source)) {
+            if (update) {
+                Files.writeString(path, newSource);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    public record Result(List<Path> files, List<Path> unsorted) {
+    }
+
+    /// Processes the C++ source files in `paths` to check if their include statements are sorted.
+    /// Include statements with any non-space characters after the closing `"` or `>` will not
```

Comment:
Perhaps this should be mentioned in the style guide?

Reply:
Done.